### PR TITLE
Consolidate unit handling in QgsUnitTypes

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR implements a number of QGIS 3.0 TODOs relating to units:

First, the distance units have been moved out of QGis and into QgsUnitTypes. This involves:
- QGis::DecimalDegrees, DegreesMinutesSeconds and DegreesDecimalMinutes have been removed, and have all been replaced with the general Degrees unit
- The distance unit types QGis::UnitType (including QGis::Meters, QGis::Feet, QGis::Degrees, QGis::NauticalMiles, QGis::Kilometers, QGis::Yards, QGis::Miles and QGis::UnknownUnit have been moved to QgsUnitTypes::DistanceUnit.
- Additionally, UnknownUnit was renamed to UnknownDistanceUnit. All methods which accepted QGis::UnitType parameters have been updated to take QgsUnitTypes::DistanceUnit instead
- The unit handling methods toLiteral, fromLiteral, tr, fromTr, fromUnitToUnitFactor have been removed from QGis. Their corresponding counterparts in QgsUnitTypes should be used instead.

Secondly, the QgsSymbolV2::OutputUnit enum was moved to QgsUnitypes::RenderUnit. This involves:
- The OutputUnit enum, including QgsSymbolV2::MM, QgsSymbolV2::MapUnit, QgsSymbolV2::Mixed, QgsSymbolV2::Pixel and QgsSymbolV2::Percentage has been moved to QgsUnitTypes and renamed to RenderUnit. QgsSymbolV2::OutputUnitList was renamed to QgsUnitTypes::RenderUnitList. All methods which previously accepted QgsSymbolV2::OutputUnit parameters or QgsSymbolV2::OutputUnitList parameters now take QgsUnitTypes::RenderUnit or QgsUnitTypes::RenderUnitList parameters respectively
- QgsSymbolLayerV2Utils::encodeOutputUnit() and decodeOutputUnit() were removed. QgsUnitTypes::encodeUnit() and QgsUnitTypes::decodeRenderUnit() should be used instead

The motivation behind this is to consolidate all the handling of different unit types into the one (well tested) class. I'd also like to make greater use of QgsSymbolV2::OutputUnit (now QgsUnitTypes::RenderUnit) outside of symbology, eg to replace the near duplicate enums in labeling. So it makes sense for this enum to be moved outside of the symbol context too.

### CODE REVIEW

**Comment by wonder-sk** [comment]
Some random thoughts:
- maybe worth having "Distance" prefix for DistanceUnit enum, as other enums have a prefix too. Right now there is Degrees value for distance and AngleDegrees for angle - these may be mixed up especially in Python where these are just plain integers
- RenderUnknownUnit instead of UnknownRenderUnit to keep the same prefix for all values (more Qt style?). Maybe unknown units could be the first value (i.e. zero), when new entries are added later, the block of "valid" values is not interrupted with "unknown" value
- maybe these enums could be in a namespace with shorter name (e.g. Qgs / QgsUnit) and only the functions would be in QgsUnitUtils or QgsUnitTypes class...?

File: python/core/qgsunittypes.sip [hunk 1]
```diff
@@ -168,5 +194,23 @@ class QgsUnitTypes
...
+     */
+    static QString encodeUnit( RenderUnit unit );
+
+    /** Decodes a render unit from a string.
+     * @param string string to decode
+     * @param ok optional boolean, will be set to true if string was converted successfully
+     * @returns decoded units
+     * @see encodeUnit()
+     */
+    static RenderUnit decodeRenderUnit( const QString& string, bool *ok = 0 );
```
Comment (wonder-sk) [inline python/core/qgsunittypes.sip]: /Out/ for ok?

**Comment by wonder-sk** [comment]
Very nice consolidation!

**Comment by nyalldawson** [comment]
> maybe worth having "Distance" prefix for DistanceUnit enum, as other enums have a prefix too. Right now there is Degrees value for distance and AngleDegrees for angle - these may be mixed up especially in Python where these are just plain integers

Good idea

> RenderUnknownUnit instead of UnknownRenderUnit to keep the same prefix for all values (more Qt style?). Maybe unknown units could be the first value (i.e. zero), when new entries are added later, the block of "valid" values is not interrupted with "unknown" value

Also good idea(s)!

> maybe these enums could be in a namespace with shorter name (e.g. Qgs / QgsUnit) and only the functions would be in QgsUnitUtils or QgsUnitTypes class...?

Hmm. Maybe we should just move all these widely used enums to a enum only class, not just the unit ones? (Bit like how qt does). The biggest issue is how slow recompilation is if you change these headers, but I guess nicer api should trump faster compilation...

Otherwise I could just rename this whole class QgsUnits?

**Comment by wonder-sk** [comment]
> Maybe we should just move all these widely used enums to a enum only class, not just the unit ones? (Bit like how qt does)

Yeah I think that would be quite nice - an equivalent of "Qt" namespace...

**Comment by nyalldawson** [comment]
Ok, I've made all the requested changes expect moving the enums to a common class. That can be done later if we decide to go down that path...

**Pull request merged**